### PR TITLE
Use enhanced for-loop syntax wherever as possible

### PR DESCRIPTION
**Pull request by amCap1712**
The enhanced for-loop syntax helps eliminate unnecessary boilerplate during iteration.

### CODE REVIEW

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/block/procedure/CaseProcedure.java [hunk 1]
```diff
@@ -55,9 +55,8 @@ public CaseProcedure()
     public void value(T argument)
     {
         int localSize = this.predicateProcedures.size();
-        for (int i = 0; i < localSize; i++)
+        for (Pair<Predicate<? super T>, Procedure<? super T>> pair : this.predicateProcedures)
```
Comment (mohrezaei) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/block/procedure/CaseProcedure.java:58]: this is not the same: so-called-enhanced for loop creates garbage.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/block/procedure/ChainedProcedure.java [hunk 2]
```diff
@@ -41,9 +41,9 @@ public void addProcedure(Procedure<? super T> procedure)
     public void value(T object)
     {
         int size = this.procedures.size();
-        for (int i = 0; i < size; i++)
+        for (Procedure<? super T> procedure : this.procedures)
```
Comment (mohrezaei) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/block/procedure/ChainedProcedure.java:44]: this is not the same: so-called-enhanced for loop creates garbage.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMutableCollection.java [hunk 3]
```diff
@@ -102,9 +102,9 @@ public boolean addAllIterable(Iterable<? extends T> iterable)
         {
             List<T> list = (List<T>) iterable;
             int size = list.size();
-            for (int i = 0; i < size; i++)
+            for (T t : list)
```
Comment (mohrezaei) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMutableCollection.java:105]: this is not the same: so-called-enhanced for loop creates garbage.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/CompositeFastList.java [hunk 4]
```diff
@@ -368,9 +368,8 @@ public void add(int index, E element)
             throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + localSize);
         }
         int max = 0;
-        for (int i = 0; i < this.lists.size(); i++)
+        for (List<E> list : this.lists)
```
Comment (mohrezaei) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/CompositeFastList.java:371]: this is not the same: so-called-enhanced for loop creates garbage.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/CompositeFastList.java [hunk 5]
```diff
@@ -402,9 +401,8 @@ public int indexOf(Object o)
     {
         int offset = 0;
         int listsSize = this.lists.size();
-        for (int i = 0; i < listsSize; i++)
+        for (MutableList<E> list : this.lists)
```
Comment (mohrezaei) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/CompositeFastList.java:404]: this is not the same: so-called-enhanced for loop creates garbage. (many more... any time there is a `list.get()`)

**Review by mohrezaei — CHANGES_REQUESTED** [review]

**Review by donraab — CHANGES_REQUESTED** [review]
Thank you for the contribution @amCap1712! As @mohrezaei has commented a few times, anywhere you replace indexed access against a List with the for-each loop syntax from Java 5, you will introduce an undesirable generation of an Iterator. We have optimized against this pattern for performance reasons since the early days of the library. Anywhere you are replacing existing Iterator creations with the simplified for-loop is a good change and is most welcome. Also, you can safely use the for-loop when iterating over arrays as long as the entire array is iterated over.

**Comment by amCap1712** [comment]
Thanks for the reviews @donraab and @mohrezaei. I have updated the PR in accordance with it.

**Review by mohrezaei — APPROVED** [review]

**Review by donraab — APPROVED** [review]
LGTM. Thanks for the contribution @amCap1712 !

**Comment by prathasirisha** [comment]
Hi @amCap1712, thank you for the contribution. Apologies for the delayed response.  Can you please rebase your PR, so one of us can merge this in.

**Comment by amCap1712** [comment]
Hi! I had updated the PR. But one check fails, don't know if its connected to changes in this PR.

**Comment by prathasirisha** [comment]
Hi @amCap1712, it may not be related to your PR. Of late we are experiencing occasional build timeouts on TeamCity. Unfortunately, we can't tell if the build would be successful or not for us to proceed with the merge. 
Can you please rebase and push once more (I merged a different PR few minutes back so your branch is 2 commits behind)?  Thank you for being patient!

**Comment by donraab** [comment]
Hi @amCap1712 If you can rebase again, @prathasirisha or I will merge your PR. The builds are passing for your PR, so if anything fails after you rebase, we'll just restart any failed builds and merge after they pass. Thanks!

**Comment by amCap1712** [comment]
Apologies for the delay. I have updated the PR.

**Pull request merged**